packages/postgrest-swr/src/subscribe/use-subscription-query.ts

```diff
@@ -0,0 +1,66 @@
+import { useSWRConfig } from "swr";
+import { useEffect, useId } from "react";
+import {
+  GenericTable,
+  PostgrestSWRMutatorOpts,
+  useCacheScanner,
+  update,
+  insert,
+  remove,
+} from "../lib";
+import { SupabaseClient } from "@supabase/supabase-js";
+import { DEFAULT_SCHEMA_NAME } from "@supabase-cache-helpers/postgrest-shared";
+import { Response, PostgresChangeFilter } from "./types";
+
+function useSubscriptionQuery<T extends GenericTable, Q extends string = "*">(
+  client: SupabaseClient | null,
+  channel: string,
+  filter: PostgresChangeFilter,
+  query: Q,
+  primaryKeys: (keyof T["Row"])[],
+  opts?: Omit<PostgrestSWRMutatorOpts<T>, "schema">
+) {
+  const { mutate } = useSWRConfig();
+  const scan = useCacheScanner(filter.table, opts);
+
+  useEffect(() => {
+    if (!client) return;
+
+    const subscription = client
+      .channel(channel)
+      .on(
+        "postgres_changes",
+        { ...filter, schema: filter.schema ?? DEFAULT_SCHEMA_NAME },
```

Comment:
Hey nice work making progress here! I like the way your tests are structured, can learn something there.

Do have some feedback on the approach.

## Initial Feedback - Overfetching
I *think* the problem with this implementation is that you're subscribing to *every* change on the target table that the user would have access to.

Based on [the supbase docs here](https://supabase.com/docs/guides/realtime/postgres-changes), it looks like the way to subscribe to things by their primary key is this:

```typescript
const channel = supabase
  .channel('db-changes')
  .on(
    'postgres_changes',
    { event: 'UPDATE', schema: 'public', table: 'messages', filter: 'id=eq.1' },
    (payload) => console.log(payload)
  )
  .subscribe()
```

Imagining for a moment that our Supabase database represents github -- and that we have a table `repos` representing all github repos (many of which are public and thus accessible to all users) -- this would be an impractical set of changes for any given client to parse.

## Proposed Addition: `useQuery` subscription extension.
Instead, my original thought was that the default behavior would be to automatically collect primary keys from the results of _**any query**_ on a given table -- and subscribe to changes *specifically* on those primary keys.

The advantage of this would be that any items you had already fetched, you'd then be subscribed to realtime changes on (provided that was something you were interested in).

It may be helpful & intuitive to make this an option you could pass into the basic `useQuery` object:

```typescript
const { data, count, isValidating, mutate, error } = useQuery(
        client
          .from("contact")
          .select("id,username", { count: "exact" })
          .eq("username", "psteinroe"),
        "multiple",
        { 
            revalidateOnFocus: false, 
            // NEW CONFIG VALUE!
            subscribeToPrimaryKeys: ["id", "username"] }
      );
```

Just to talk through what I imagine would happen here:

(1) Initial data gets fetched, let's say it looks like:

```typescript
// The data we fetched during initial Revalidate
{id: "the-id", username: "psteinroe"}
```

(2) We look at the data we fetched. If the `subscribeToPrimaryKeys` config is not empty, we would subscribe to `id=whatever` and `username=psteinroe`.

(3) If these values are to *exit* the cache at any time, we keep track of this, and we unsubscribe from these supabase subscriptions at that time.


## Limitations
I understand that for situations like lists, and groups of items, this may be suboptimal -- but until we get better query capabilities coming out of Realtime-js [1], [2] -- I think this is the best we can do.


[1] https://github.com/supabase/realtime-js/issues/185
[2] https://github.com/supabase/realtime/issues/222

Reply:
I'd be happy to try my hand at adding this functionality as a PR onto this PR, if you would like.

Reply:
Hey, thanks for your elaborate feedback (and putting up with a PR with no documentation, sorry!).

Let me first add the documentation and explain my thought processes before I respond to your feedback and finally make a proposal with what I hope to be the best of both worlds.

### Documentation
The `useSubscription` hook simply manages a realtime subscription. Upon retrieval of an update, it updates the cache with the retrieved data the same way the mutation hooks do. It exposes all params of the `.on()` method, including the `filter`. Hence, your first example could be implemented as

```ts
// your example
const channel = supabase
  .channel('db-changes')
  .on(
    'postgres_changes',
    { event: 'UPDATE', schema: 'public', table: 'messages', filter: 'id=eq.1' },
    (payload) => console.log(payload)
  )
  .subscribe()

// would be
useSubscription(
         client.channel("db-changes"),
         { event: "UPDATE", table: "messages", schema: "public", filter: "id=eq.1" },
         ["id"]
       );
```

The `useSubscriptionQuery` hook does exactly the same, but instead of updating the cache with the data sent by realtime, it fetches the latest data from PostgREST and updates the cache with that. The main use case for this hook are [Computed Columns](https://postgrest.org/en/stable/api.html?highlight=computed%20columns#computed-virtual-columns), because these are not sent by realtime. 

### The Overfetching Issue
If I understand your concerns correctly, you propose to open a subscription for each entity based on their primary keys to not receive all updates of a table. While I think this works fine if you are exclusively work with `single` or `maybeSingle` queries, it can quickly lead to hundreds of subscriptions if you are using list queries. With the current implementation, your second example would be implemented by opening a single subscription that effectively uses the same filter as its query.

```ts
const { data, count } = useQuery(
         client
           .from("messages")
           .select("id,username", { count: "exact" })
           .eq("username", "myusername"),
         "multiple"
       );

useSubscription(
         client.channel("db-changes"),
         { event: "UPDATE", table: "messages", schema: "public", filter: "username =eq. myusername" },
         ["id"]
       );

// when an update is received, the "data" from useQuery would be automatically updated
```

This of course falls short if you do not set any filters on your query. e.g. in your repos example

```ts
// data would contain the first n (limit) repos
const { data, count } = useQuery(
         client
           .from("repos")
           .select("id, name", { count: "exact" }),
         "multiple"
       );

// we would receive an update for all repos in this world and prob die shortly after 
useSubscription(
         client.channel("repo-changes"),
         { event: "UPDATE", table: "repos", schema: "public" },
         ["id"]
       );
```

But in such a case I would move the realtime functionality into the repo and opening a subscription for each, e.g.

```ts
// data would contain the first n (limit) repos
const { data, count } = useQuery(
         client
           .from("repos")
           .select("id, name", { count: "exact" }),
         "multiple"
       );

for repo in data...
    render the repo list item and within the list item, do

    useSubscription(
             client.channel("repo-changes"),
             { event: "UPDATE", table: "repos", schema: "public" , filter="id.eq.myrepoid"},
             ["id"]
           );
```

As far as I understand, your proposal would allow for a different DX in the case where I have a large unfiltered list and want to receive updates for all items that were queried from that list. My main concerns here are that adding it to the query would add even more "magic" to the lib, and reduce the separation of concern since the same functionality can be achieved by using `useQuery` with `useSubscription`.

### Best of both worlds (I hope)
First of all, I totally agree with the both issues you mentioned. The current DX around realtime-is not nice. Especially the untyped client and limited filters should be prioritised. Maybe we can help here a bit by providing a helper that turns a PostgREST query into a realtime filter (as good as it is possible).

```ts
const query = client
           .from("messages")
           .select("id,username", { count: "exact" })
           .eq("username", "myusername")
const { data, count } = useQuery(
         query,
         "multiple"
       );

// subscribe to the smallest amount of updates that include all results of a query, in this case username.eq.myusername
useSubscription( 
         client.channel("db-changes"),
         { event: "UPDATE", table: "messages", schema: "public", filter: fromQuery(query) },
         ["id"]
       );
```

Looking forward to your feedback!

Reply:
btw I have no idea why the test fail. I tried to debug it but it seems like realtime just does not work. The subscription is not created. I even stumbled upon [your issue](https://github.com/supabase/supabase/issues/6290), but I cannot see anything wrong with the current setup.